oauthproxy.go

```diff
@@ -58,17 +58,18 @@ type allowedRoute struct {
 
 // OAuthProxy is the main authentication proxy
 type OAuthProxy struct {
-	CookieSeed     string
-	CookieName     string
-	CSRFCookieName string
-	CookieDomains  []string
-	CookiePath     string
-	CookieSecure   bool
-	CookieHTTPOnly bool
-	CookieExpire   time.Duration
-	CookieRefresh  time.Duration
-	CookieSameSite string
-	Validator      func(string) bool
+	CookieSeed           string
+	CookieName           string
+	CSRFCookieName       string
+	CookieDomains        []string
+	CookiePath           string
+	CookieSecure         bool
+	CookieHTTPOnly       bool
+	CookieExpire         time.Duration
+	CookieRefresh        bool
```

Comment:
Why do we make this a bool rather than a duration? Some users may want to make this an explicit duration and not use the percentage option, this will break those existing users

Reply:
For user simplicity, how do we feel about overloading this option to take in a plain number of seconds OR a percent (based on a trailing `%` in the settings) and making this a `string` (and doing the appropriate validation and conversion next).

https://github.com/oauth2-proxy/oauth2-proxy/pull/886#discussion_r516455795

Also long term, we should rename this (v8 structured configs?). This is more accurately a session refresh, and now we have non-cookie based sessions (redis and soon AWS dynamodb).

Reply:
> Why do we make this a bool rather than a duration? Some users may want to make this an explicit duration and not use the percentage option, this will break those existing users

Because if refresh is disabled no refresh happens. I preserved the ability to not have any refresh and let to cookie expire unmodified regardless of contents.

> For user simplicity, how do we feel about overloading this option to take in a plain number of seconds OR a percent (based on a trailing `%` in the settings) and making this a `string` (and doing the appropriate validation and conversion next).

You have to call the option something else then -  "-percent" cannot be an absolute value *and* a percent and still be semantically correct. If you come up with a name for an abstraction that can accept both an arbitrary absolute value in seconds and a percentage of some externally defined arbitrary value, I'll be happy to implement it.

Reply:
Another important point I'd like to raise is that from a practical standpoint **fixed seconds refresh has no actual value**. 

The IdP sets token expiration based on policies that can change literally from refresh to refresh based on IdP behavior alone. Having a session that carries a token that has expired because your fixed refresh seconds don't match whatever IdP set for access token and id_token defeats the entire point of this proxy. That's why the issue of #856 even popped up in the first place - having a session with expired tokens is pointless, session's only value is in the usable tokens it carries. 

Furthermore, different users coming through the proxy via the *same* IdP may have the different token expiration (again, due to dynamic policies) so the fixed seconds may not even work at all for all users. 

Percentage works regardless of IdP settings. Fixed expiration almost never works.

Reply:
> You have to call the option something else then - "-percent" cannot be an absolute value and a percent and still be semantically correct. If you come up with a name for an abstraction that can accept both an arbitrary absolute value in seconds and a percentage of some externally defined arbitrary value, I'll be happy to implement it.

Yeah, I'm suggesting we ditch any new or redundant flags and just keep using `cookie-refresh` for both cases.

Otherwise we need to handle validation & startup errors if both values are set and how to deconflict them.  Likely much easier and user friendly to handle `seconds` or `percentage` in a single config option.

As for flag names, I prefer brevity over semantic correctness if the latter leads to overly long configuration flag names.

Like I mentioned above, we don't need to split hairs too much about the accuracy of the naming of `cookie-refresh*` since it is already wrong in the name for what it functionally does and needs a rename in v8.

Reply:
> The IdP sets token expiration based on policies that can change literally from refresh to refresh based on IdP behavior alone.

I think alot of users have IdPs/Providers where we can customize the IDToken TTL exactly to our specification. And for those users, the hardcoded time is fine since they get absolute control to align the timing between the two values.

But good point on some IdPs not allowing that to be configured. Hopefully it is relatively static, but I can see it changing being annoying.

Reply:
The question I would have is why then? A setting of "100%", although useless (due to being corner case of token expiration), is equivalent to "whatever IdP set the token expiration to be". And my comment didn't imply that users couldn't set their IdP expiration but rather that those settings could be dynamic and flexible.

To ask the same question in a different way - what is a use case having to set a specific refresh time? I set it to 10 minutes, admin of the IdP (who is not me) reconfigures it to be 5 minutes, users start getting errors, nobody knows why. Why would anyone want to set a specific refresh period that is functionally UNRELATED to IdP token expiration times?

Reply:
Backwards compatibility - all existing users have a static time. I agree percent is a much nicer experience. But the v7 upgrade should be a drop-in replacement for most users for the most part just updating a docker image tag and not require our whole user base to change a flag value. 

The percent concept is a little bit of a scope creep (while nice) from the actual bug that needs to be fixed that is causing issues: We need to allow refreshing sessions before a token is expired IF the user configures that to happen (with a `cookie-refresh` shorter than the token TTL or percent less than 100%, whichever).